### ISSUE TITLE
dealing with timeouts?

### ISSUE THREAD

**Issue by sstephans** [geopy/geopy#344]
So, when running my function, I'm getting the following timeouts:

```
Traceback (most recent call last):
  File "C:\Python37\lib\urllib\request.py", line 1317, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "C:\Python37\lib\http\client.py", line 1229, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "C:\Python37\lib\http\client.py", line 1275, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "C:\Python37\lib\http\client.py", line 1224, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "C:\Python37\lib\http\client.py", line 1016, in _send_output
    self.send(msg)
  File "C:\Python37\lib\http\client.py", line 956, in send
    self.connect()
  File "C:\Python37\lib\http\client.py", line 1392, in connect
    server_hostname=server_hostname)
  File "C:\Python37\lib\ssl.py", line 412, in wrap_socket
    session=session
  File "C:\Python37\lib\ssl.py", line 853, in _create
    self.do_handshake()
  File "C:\Python37\lib\ssl.py", line 1117, in do_handshake
    self._sslobj.do_handshake()
socket.timeout: _ssl.c:1039: The handshake operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python37\lib\site-packages\geopy\geocoders\base.py", line 344, in _call_geocoder
    page = requester(req, timeout=timeout, **kwargs)
  File "C:\Python37\lib\urllib\request.py", line 525, in open
    response = self._open(req, data)
  File "C:\Python37\lib\urllib\request.py", line 543, in _open
    '_open', req)
  File "C:\Python37\lib\urllib\request.py", line 503, in _call_chain
    result = func(*args)
  File "C:\Python37\lib\urllib\request.py", line 1360, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "C:\Python37\lib\urllib\request.py", line 1319, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error _ssl.c:1039: The handshake operation timed out>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\saman\Documents\ancestry\python-gedom-playground\markersfix.py", line 49, in <module>
    geocode_temp = geolocator.geocode(place, timeout=None)
  File "C:\Python37\lib\site-packages\geopy\geocoders\osm.py", line 309, in geocode
    self._call_geocoder(url, timeout=timeout), exactly_one
  File "C:\Python37\lib\site-packages\geopy\geocoders\base.py", line 367, in _call_geocoder
    raise GeocoderTimedOut('Service timed out')
geopy.exc.GeocoderTimedOut: Service timed out
```

I'm using the library to get latitudes and longitudes of places from strings. I included `from geopy.exc import GeocoderTimedOut` and in the single call to geocode() added this argument `geocode_temp = geolocator.geocode(place, timeout=None)` 

I also ensured that for duplicate strings, the program is not making multiple calls in a row to geocode(). Oh, and the timeout is not happening on the same entry, in other words, sometimes it times out after the second call, sometimes after the 15th, etc. It's not reproducible based on what is being searched for with geocode().

The *really* strange thing is that the exact same program was working just fine yesterday, albeit on my school's internet network, and now I'm at home (though I can't imagine why that should matter). I looked on my google API dashboard and nothing is jumping out at me like I've reached usage limits or anything.

**Comment by KostyaEsmukov** [comment]
Apparently there's a confusion in the docs, saying that `None` means "no timeout". Actually this is true for `geopy.geocoders.options.default_timeout` only. For timeout set as a method arg this is not true: `None` there means "use default timeout, as set by `geopy.geocoders.options.default_timeout`". If you run python with warnings enabled (the `-Wd` switch), you should get a warning:

https://github.com/geopy/geopy/blob/ff71a6dff80fd7f65ba3820411d9da363b4587ff/geopy/geocoders/base.py#L330-L338

To workaround the issue, you may move that None to `geopy.geocoders.options.default_timeout` instead, or just pass a large number as a timeout (which seems to be a better option anyways, as infinite timeout for I/O operations doesn't seem to make a lot of sense).

The networking issues are really out of the control of geopy, and are probably arised due to bulk geocoding with Nominatim (which you seem to be using, according to the provided traceback). Nominatim, actually, discourages usage of their service for that in their ToS: https://operations.osmfoundation.org/policies/nominatim/#bulk-geocoding

If you're trying to do bulk lookups, take a look at the `geopy.extra.rate_limiter.RateLimiter` class: https://geopy.readthedocs.io/en/1.18.1/#usage-with-pandas

If the dataset is large, you should consider switching to another geocoding service.

**Comment by KostyaEsmukov** [comment]
> So with that in mind, can you recommend a python library that is better suited to this kind of bulk lookup? 

It's not the library you should be thinking of, it's the geocoding service(s) you would use.

> the numbers can go into the low thousands

For Nominatim it is a lot, but for other commercial options it should be fine. For example, Google provides $200 worth of credits on geocoding for free each month, that should be enough for ~40.000 requests for free. https://developers.google.com/maps/documentation/geocoding/usage-and-billing

Closing, as this is really a usage question rather than a bug.

**Comment by KostyaEsmukov** [comment]
The docs confusion has been fixed in 20afeaf921f161cbbc330da7bf6029f82b87f0f8
